### PR TITLE
Produce a combined coverage report when running Go unit tests, and use goveralls to report coverage results from Travis to Coveralls.io.

### LINKED ISSUE #4510: Unit test coverage report dashboard
Right now, we execute unit tests in Travis with `-cover`, and you can see the results on each line:

```
ok  	github.com/GoogleCloudPlatform/kubernetes/pkg/api	19.821s	coverage: 41.7% of statements
```

but we don't have an easy place to go browse our code and figure out what's covered and what isn't. Is there some awesome way to do this? At the very least, can we get the actual line-by-line coverage results integrated into our CI build process so we can poke at these somehow?

### PR DESCRIPTION
**Pull request by ixdy**
Fixes #4510.

I've tested the hack/test-go.sh part of this change in a variety of ways by running
- hack/test-go.sh ""
- hack/test-go.sh dir1/ dir2/ etc
- both of the above with KUBE_COVER=y
- both of the above with KUBE_GOVERALLS_BIN set (though it fails, possibly because it's not running from Travis?)
- hack/test-integration.sh
- hack/benchmark-go.sh]
- build/release.sh.

I have no idea how to test the Travis changes. I guess in the worst case we just won't get any coverage results uploaded, which is no worse than before.
At least running locally now gives a nice HTML report.

### CODE REVIEW

**Comment by ixdy** [comment]
OK, that answers the question on whether it works. Awesome!

**Comment by zmerlynn** [comment]
Do you know why the source links aren't working on https://coveralls.io/builds/1953672, like https://coveralls.io/builds/1953672/source?filename=pkg%2Fclient%2Fconditions.go>

**Comment by ixdy** [comment]
That's weird - it definitely worked yesterday. I wonder if the kubernetes -> kubernetes_copy -> kubernetes switch caused something to break.

I'm kicking off the Travis build again for this PR to see if it fixes it.

**Comment by zmerlynn** [comment]
This LGTM overall, but there's enough bash here that I'm cc'ing @filbranden.

**Comment by ixdy** [comment]
Coveralls is still not showing the code, though it looks like they had a new feature push overnight - maybe something broke? I'll follow up with a support request there.

**Comment by ixdy** [comment]
Integration is working properly in #4629 (e.g. https://coveralls.io/builds/1960240/source?filename=pkg%2Fclient%2Fconditions.go). From talking to Coveralls, they are querying GitHub with a commit ref that GitHub no longer thinks exists; it's possible that something got invalidated by the repo rename.

**Comment by zmerlynn** [comment]
Nice. LGTM.

File: hack/test-go.sh [hunk 1]
```diff
@@ -135,33 +136,50 @@ if [[ $iterations -gt 1 ]]; then
...
 fi
 
+cover_report_dir=""
+combined_cover_profile=""
+if [[ -n "${KUBE_COVER}" ]]; then
+  cover_report_dir="/tmp/k8s_coverage/$(kube::util::sortable_date)"
+  combined_cover_profile="${cover_report_dir}/combined-coverage.out"
+  kube::log::status "Saving coverage output in '${cover_report_dir}'"
+  mkdir -p ${cover_report_dir}
+  echo "mode: ${KUBE_COVERMODE}" >${combined_cover_profile}
```
Comment (zmerlynn) [inline hack/test-go.sh]: Can you comment what's going on with the `mode:`? I worked it out but it would be kind to the next reader.

Comment (ixdy) [inline hack/test-go.sh]: Done.

**Comment by ddysher** [comment]
Is it me or does anyone observes test-go.sh becomes much slower after the PR?    I think xargs will spawn multiple process, so we have a faster execution.

**Comment by ixdy** [comment]
The xargs call would result in only a single "go test" command, I think, but it's likely that Go internally parallelized things. We are calling it sequentially now, yes.

Looking at recent Travis builds, it looks like the test-go.sh execution time increased from ~60s to ~9-10m, which is a pretty bad regression. I'll see what can be done to improve that.

**Comment by bgrant0607** [comment]
Total test time has more than tripled. Travis is getting farther and farther behind, particularly on pull requests. If this can't be sped up, do we really need coverage on every PR and every build?

**Comment by ixdy** [comment]
Working on a PR, should be improved later today hopefully.

**Comment by bgrant0607** [comment]
What is the fix in progress?

Is there a way to disable this for PRs? Builds are frequently taking more than 25 minutes.

**Comment by ixdy** [comment]
PR is the one referenced on Friday - #4686.

**Comment by filbranden** [comment]
If this is hurting productivity we should just revert the offending PR... no?

**Comment by ixdy** [comment]
Sure, let me do that.

**Pull request merged**